Add ScriptStateListener for easier Script debugging.

shadders opened the pull request:
ScriptStateListener allows peeking inside a Script during execution. It's intended as a basis for both debugging and with a nice wrapper could be used as an educational tool.

I have made notes in the class JavaDoc about the lack of immutability of the script stacks:

> Note that the stacks are set as local variables in the setup stage as unModifiable lists.  However they are backed by the live list in the script engine and as such will update on each call to the listener.  Also important to note that although the lists themselves cannot be modified the elements are byte arrays and as such could be modified.  Here be dragons!
> 
> This class is not recommended for real world use.  Only for testing and debugging scripts.

By doing it this way the script state is set in the listener once on setup and save having to pass a large number of arguments on each call.  An alternative would be to explicitly copy the stack/altStack on each call to the listener.  This would mean cloning the byte[] elements of the lists as well.  Slower and more cumbersome to implement (but if the scriptStateListener is null that's all skipped).  But would ensure there's no way the listener can modify the internal state of the script interpreter.

It may in fact be desirable to allow this mutability however.  A script debugger may want to allow you modify, add or remove stack elements and even op codes during operation similar to how IDEs do.  Open for discussion on this point.

shadders committed:
Add ScriptStateListener for easier Script debugging.

shadders committed:
removed mystery import

HashEngineering commented:
This looks interesting.  In the past, when I was working on the Signature Hash for my bitcoinj fork for BCH, I had a lot of trouble trying to figure out what was going on with the Script that contained the signatures.  Initially, my signature hash method was written from scratch and had some bugs.  Later, I found out the signature hash was the same as with Bitcoin Segwit transactions.  Nevertheless, this pull request may prove to be useful with smart contracts or other scripts that are more complicated than P2PKH.

shadders committed:
move InteractiveScriptStateListener to tools project.
rename onScriptComplete() to onScriptSuccess()
use a working pair of scriptSig and scriptPubKey in the example.

shadders committed:
fix private/protected access errors

shadders commented:
For reference here's a sample output from the demo implementation.

The script actually fails due the unsatisfied link error although it is actually a valid scriptPubKey and scriptSig pair.

```org.bitcoinj.tools.InteractiveScriptStateListener

***Executing scriptSig***

PUSHDATA(71)[3044...8301] PUSHDATA(33)[02e5...b9a4]

Executing PUSHDATA operation: [NON_OP(71)]
Execution point (^): ^PUSHDATA(71)[3044...8301] PUSHDATA(33)[02e5...b9a4]

Stack:
index[0] len[71] [304402201c999cf44dc6576783c0f55b8ff836a1e22db87ed67dc3c39515a6676cfb58e902200b4a925f9c8d6895beed841db135051f8664ab349f2e3ea9f8523a6f47f9388301]


Executing PUSHDATA operation: [NON_OP(33)]
Execution point (^): PUSHDATA(71)[3044...8301] ^PUSHDATA(33)[02e5...b9a4]

Stack:
index[0] len[33] [02e58d7b931b5d43780fda0abc50cfd568fcc26fb7da6a71591a43ac8e0738b9a4]
index[1] len[71] [304402201c999cf44dc6576783c0f55b8ff836a1e22db87ed67dc3c39515a6676cfb58e902200b4a925f9c8d6895beed841db135051f8664ab349f2e3ea9f8523a6f47f9388301]

Script success.

***Executing scriptPubKey***

DUP HASH160 PUSHDATA(20)[f7d5...ccdc] EQUALVERIFY CHECKSIG

Executing OP_CODE operation: [DUP]
Execution point (^): ^DUP HASH160 PUSHDATA(20)[f7d5...ccdc] EQUALVERIFY CHECKSIG

Stack:
index[0] len[33] [02e58d7b931b5d43780fda0abc50cfd568fcc26fb7da6a71591a43ac8e0738b9a4]
index[1] len[33] [02e58d7b931b5d43780fda0abc50cfd568fcc26fb7da6a71591a43ac8e0738b9a4]
index[2] len[71] [304402201c999cf44dc6576783c0f55b8ff836a1e22db87ed67dc3c39515a6676cfb58e902200b4a925f9c8d6895beed841db135051f8664ab349f2e3ea9f8523a6f47f9388301]


Executing OP_CODE operation: [HASH160]
Execution point (^): DUP ^HASH160 PUSHDATA(20)[f7d5...ccdc] EQUALVERIFY CHECKSIG

Stack:
index[0] len[20] [f7d52018971f4ab9b56f0036958f84ae0325ccdc]
index[1] len[33] [02e58d7b931b5d43780fda0abc50cfd568fcc26fb7da6a71591a43ac8e0738b9a4]
index[2] len[71] [304402201c999cf44dc6576783c0f55b8ff836a1e22db87ed67dc3c39515a6676cfb58e902200b4a925f9c8d6895beed841db135051f8664ab349f2e3ea9f8523a6f47f9388301]


Executing PUSHDATA operation: [NON_OP(20)]
Execution point (^): DUP HASH160 ^PUSHDATA(20)[f7d5...ccdc] EQUALVERIFY CHECKSIG

Stack:
index[0] len[20] [f7d52018971f4ab9b56f0036958f84ae0325ccdc]
index[1] len[20] [f7d52018971f4ab9b56f0036958f84ae0325ccdc]
index[2] len[33] [02e58d7b931b5d43780fda0abc50cfd568fcc26fb7da6a71591a43ac8e0738b9a4]
index[3] len[71] [304402201c999cf44dc6576783c0f55b8ff836a1e22db87ed67dc3c39515a6676cfb58e902200b4a925f9c8d6895beed841db135051f8664ab349f2e3ea9f8523a6f47f9388301]


Executing OP_CODE operation: [EQUALVERIFY]
Execution point (^): DUP HASH160 PUSHDATA(20)[f7d5...ccdc] ^EQUALVERIFY CHECKSIG

Stack:
index[0] len[33] [02e58d7b931b5d43780fda0abc50cfd568fcc26fb7da6a71591a43ac8e0738b9a4]
index[1] len[71] [304402201c999cf44dc6576783c0f55b8ff836a1e22db87ed67dc3c39515a6676cfb58e902200b4a925f9c8d6895beed841db135051f8664ab349f2e3ea9f8523a6f47f9388301]


Executing OP_CODE operation: [CHECKSIG]
Feb 07, 2018 6:10:45 PM org.bitcoin.Secp256k1Context <clinit>
INFO: java.lang.UnsatisfiedLinkError: no secp256k1 in java.library.path
Execution point (^): DUP HASH160 PUSHDATA(20)[f7d5...ccdc] EQUALVERIFY ^CHECKSIG

Stack:
index[0] len[0] []

Script failed.

Process finished with exit code 0
```

shadders committed:
add test transaction source URLs

shadders committed:
add script parser
